core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -160,7 +171,18 @@ public AllocationService(Settings settings, AllocationDeciders allocationDecider
         if (!reroute(allocation)) {
             return new RoutingAllocation.Result(false, clusterState.routingTable());
         }
-        return new RoutingAllocation.Result(true, new RoutingTable.Builder().updateNodes(routingNodes).build().validateRaiseException(clusterState.metaData()));
+        RoutingAllocation.Result result = new RoutingAllocation.Result(true,
+                new RoutingTable.Builder().updateNodes(routingNodes).build().validateRaiseException(clusterState.metaData()));
+        logClusterHealthStateChange(clusterState, ClusterState.builder(clusterState).routingResult(result).build(), "rerouting");
+        return result;
+    }
+
+    private void logClusterHealthStateChange(ClusterState previousState, ClusterState newState, String reasonForChange) {
+        ClusterHealthStatus previousHealth = new ClusterHealth(previousState).getStatus();
+        ClusterHealthStatus currentHealth = new ClusterHealth(newState).getStatus();
+        if (!previousHealth.equals(currentHealth)) {
+            logger.info("Cluster health status changed from {} to {} when {}.", previousHealth, currentHealth, reasonForChange);
```

Comment:
we typically log the changing parts on `[]`. Like `cluster health changed from [{}] to [{}] (reason [{}])`. wdyt?